nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java

```diff
@@ -3122,70 +3356,150 @@ public void createGradFunction() {
+                    Preconditions.checkState(variables.containsKey(s), "Specified loss function variable \"%s\" does not exist", s);
+                    SDVariable v = variables.get(s).getVariable();
+                    Preconditions.checkState(v.dataType().isFPType(), "Specified loss function variable \"%s\" is not a floating" +
+                            "point variable (datatype: %s). Only floating point variables may be used as loss function variable", s, v.dataType());
+                    v = v.sum();    //If output is not a scalar: we'll use loss = v.sum(), same as adding loss for multiple outputs. We don't always know for sure if output is scalar at this point
+                    if(v.dataType() == initialGrad.dataType()){
+                        sameDiff.setGradientForVariableName(v.getVarName(), initialGrad);
+                    } else {
+                        sameDiff.setGradientForVariableName(v.getVarName(), initialGrad.castTo(v.dataType()));
                     }
-                    if(!isExternalGrad && (v.getVariable().getVariableType() != VariableType.ARRAY ||
-                            (v.getInputsForOp() != null && ! v.getInputsForOp().isEmpty())
-                            || !v.getVariable().dataType().isFPType())){
-                        continue;
+                    if(finalOutputs.contains(v)){
+                        log.warn("Loss function variable \"{}\" appears multiple times in list of loss variables - using only first instance", s);
+                    } else {
+                        finalOutputs.add(v);
                     }
-                    finalOutputs.add(v.getVariable());
                 }
 
-                Preconditions.checkState(!finalOutputs.isEmpty(), "Could not infer final network outputs to begin differentiation");
-
                 if (log.isTraceEnabled()) {
                     String[] initialOutputsStr = allFunctions.get(allFunctions.size() - 1).getOp().outputVariablesNames();
                     String s = initialOutputsStr == null ? "null" : Arrays.toString(initialOutputsStr);
                     log.trace("Defining backward function: initial outputs {}", s);
                 }
 
-                //Differentiate ops in any valid reverse topological order to ensure the parent gradient nodes are
-                // available before we try to differentiate the op
-                Queue<DifferentialFunction> availableForDiff = new LinkedList<>();
-                Set<String> seenOps = new HashSet<>();
-                //start with scalar backprop
-                INDArray initGradArr;
-                try(MemoryWorkspace ws = Nd4j.getMemoryManager().scopeOutOfWorkspaces()) {
-                    initGradArr = Nd4j.scalar(1.0);
+
+                //----- Step 1: Determine FP variables connected to loss -----
+                // Find all FP variables that are connected to loss by an FP32 path
```

Comment:
Should this actually be only an fp32 path?

Reply:
No, any FP is fine. Good catch.